app/controllers/votes_controller.rb

```diff
@@ -0,0 +1,22 @@
+class VotesController < ApplicationController
+  before_action :require_authentication
+
+  before_action :find_or_initialize_vote, only: %i[create destroy]
+
+  def create
+    redirect_to course_path(@vote.course_slug)
+  end
+
+  def destroy
+    @vote.destroy
+    redirect_to course_path(@vote.course_slug)
+  end
+
+  private
+
+  def find_or_initialize_vote
+    @vote = current_user.votes.find_or_initialize_by(course_slug: params[:course_id])
+    @vote.user_id = current_user.id
+    @vote.save
```

Comment:
The `.save` should happen in the `create` action and this method should only find or _initiliaze_, otherwise in cases where `destroy` is hit for a non-existing vote it will temporarily create one in the db.

Then in the `destroy` action you'll want to do:

```ruby
@vote.destroy if @vote.persisted?
```

Also: should use `.save!` so an exception is raised.

Reply:
Done.